src/Command/VarnishControllerDeployCommand.php

```diff
@@ -0,0 +1,170 @@
+        $verifyTLS = $this->convertToBoolean($this->requireStringOption($input, 'verify-tls'));
+
+        $client = new VarnishControllerClient($uri, $username, $password, $verifyTLS);
+        $rollbackID = '';
+
+
+        /**
+         * 1. get VCL ID by name and the content (source)
+         * 2. check if input $vclInput is set
+         * 3. check if input $vclInput is same as source from call in point 1.
+         * 3.1 if it's different get a rollbackId
+         * 3.2 update VCL with newVCL
+         * 4.
+         */
+
+        $groupID = $client->getGroupID($vclGroup);
+
+        #region start
+        $vclFile = new VclFile(
+            $client->getVclIdAndSource($vclName)
+        );
+        #endregion
+
+
+
+
+
+        if (!$vclFile->getId()) {
+            $vclFile->fillFromPayload(
+                $client->createEmptyVCL($vclName,$groupID)
+            );
+            $logger->log('info', 'Created default VCL for {name} as ID {id}', ['name' => $vclName, 'id' => $vclFile->getId()]);
+        }
+
+        $newVclSource = $this->readFile($vclInput);
+        if (!$newVclSource) {
+            throw new \InvalidArgumentException(sprintf('Filename "%s" is not readable.', $vclInput));
+        }
+        if ($newVclSource === $vclFile->getSource()) {
+            $logger->log('info', 'VCL did not change, will not update');
+        } else {
+            //todo
+            $rollbackID = $client->getHead($vclFile->getId());
+            $logger->log('info', 'Using VCL {id} which head is {rollbackID}', ['id' => $vclFile->getId(), 'rollbackID' => $rollbackID]);
+
+            $vclCommit = $client->updateVCL($vclFile->getId(), $newVclSource);
+            $logger->log('info', 'Pushed new VCL with the commit ID {id}', ['id' => $vclCommit]);
+        }
+
+
+        if (!$groupID) {
+            throw new InvalidOptionException($vclGroup.' needs to be a valid group name');
+        }
+
+        $logger->log(
+            'info', 'Deploying VCL {id} to group {groupID}', ['id' => $vclID, 'groupID' => $groupID]);
+        [$success, $compilationData, $deployData] = $client->deploy($groupID, $vclID);
+        if (!$success) {
+            $logger->log('error', 'Deployment failed');
+            $logger->log('error', 'Compilation errors: {errors}', ['errors' => Json::encode($compilationData)]);
+            $logger->log('error', 'Deployment errors: {errors}', ['errors' => Json::encode($deployData)]);
+
+            $id = $client->rollback($vclID, $rollbackID);
+            $logger->log('info', 'Rolled VCL {vclID} back to {id}', ['vclID' => $vclID, 'id' => $id]);
+
+            return 1;
+        }
+
+        $logger->log('info', 'Successfully deployed VCL {id} to group {groupID}', ['id' => $vclID, 'groupID' => $groupID]);
+        $logger->log('info', 'Compilation data: {data}', ['data' => Json::encode($compilationData)]);
+        $logger->log('info', 'Deployment data: {data}', ['data' => Json::encode($deployData)]);
```

Comment:
i would only print the data with `debug` verbosity, if at all.

why do we have two different "data"? can we find better names to make this more readable?

Reply:
Hi David, thanks for the feedback, this and the @file_get_contents was 1:1 copied from the class before, i agree there's stuff that could be improved, but i rather focused on the API stuff. As i created a new File, it's more difficult to see the differences, i could also pack the changes in the already existing "VacClient.php" so it's easier to see the changes that were made?

Also this is an intermediate state for Emanuele, there's still the thing with "head" and "rollback" ongoing, which i am not sure how to solve.

Reply:
i think keeping the data at debug verbosity could still be useful.

it depends on the overlap vs differences to know if its better to share code or have 2 separate things. if its 80% the same, it is easier to maintain if we share code. if its rather like 50% different, i find it better to have the 2 implementations not share code - or extract common functionality into a helper used by both.